safehttp/plugins/xsrf/xsrf.go

```diff
@@ -34,40 +35,64 @@ type UserIdentifier interface {
 }
 
 // Interceptor implements XSRF protection. It requires an application key and a
-// storage service. The appKey uniquely identifies each registered service and
+// storage service. The AppKey uniquely identifies each registered service and
 // should have high entropy. The storage service supports retrieving ID's of the
-// application's users. Both the appKey and user ID are used in the XSRF
-// token generation algorithm.
+// application's users. Both the AppKey and user ID are used in the XSRF
+// token generation and verification algorithm.
 type Interceptor struct {
 	AppKey     string
 	Identifier UserIdentifier
 }
 
+type tokenCtxKey struct{}
+
+// Token returns the XSRF token from the safehttp.IncomingRequest Context, if
+// present, and, otherwise, returns an empty string.
+func Token(r *safehttp.IncomingRequest) string {
+	tok := r.Context().Value(tokenCtxKey{})
+	if tok == nil {
+		return ""
+	}
+	return tok.(string)
+}
+
 // Before should be executed before directing the safehttp.IncomingRequest to
-// the handler to ensure it is not part of the Cross Site Request
-// Forgery. It checks for the presence of an xsrf-token in the request body and
-// validates it based on the userID associated with the request.
-func (p *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
-	userID, err := p.Identifier.UserID(r)
+// the handler to ensure it is not part of the Cross-Site Request
+// Forgery. In case of state changing methods, it checks for the
+// presence of an xsrf-token in the request body and validates it based on the
+// userID associated with the request. It also adds a cryptographically safe
+// XSRF token to the safehttp.IncomingRequest context.
+func (i *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	userID, err := i.Identifier.UserID(r)
 	if err != nil {
 		return w.ClientError(safehttp.StatusUnauthorized)
 	}
-	f, err := r.PostForm()
-	if err != nil {
-		mf, err := r.MultipartForm(32 << 20)
+
+	tok := xsrftoken.Generate(i.AppKey, userID, r.URL.String())
+	r.SetContext(context.WithValue(r.Context(), tokenCtxKey{}, tok))
```

Comment:
This is entirely optional but maybe we should move the generation of the new token after the validation of the old token. This way you can't accidentally in some code change in the future validate the new token instead of the old one. It might sound crazy, but I mean, bugs happen. :p

Reply:
Wholeheartedly agree with this sentiment. @mihalimara22, could you do this?